### PR TITLE
DSv4: implement decode MoE router orchestration body

### PR DESCRIPTION
**Pull request by zhaozhaozz**
## Summary
- Fill the DSv4 decode MoE router orchestration body by composing the shared hc_pre FFN stage and router kernel path.
- Add RMSNorm, gate score computation, top-k extraction, weight normalization, and scatter into router indices/weights outputs.
- Promote the router file to `models/deepseek/v4/deepseek_v4_decode_moe_router.py` and keep comments focused on the current implementation state.

## Tests
- `pre-commit run --show-diff-on-failure --color=always --all-files`
- `pytest tests/golden -q`
- `PYTHONPATH=$PWD PTOAS_ROOT=/home/bot/code/PTO/ptoas-bin PTO_ISA_ROOT=/home/bot/code/PTO/pto-isa ./.venv/bin/python models/deepseek/v4/deepseek_v4_decode_moe_router.py -p a2a3sim`
- `/home/bot/code/PTO/local-github-ci/local-ci --repo-root "$(pwd)" --job pre-commit` reached passing pre-commit checks; local act post-cache save did not exit, so the wrapper process was stopped after the check output passed.

### CODE REVIEW

**Comment by zhaozhaozz** [comment]
Follow-up on the unused `tid2eid` / `input_ids` review: I rechecked the public contract against `models/deepseek/v4/deepseek_v4_decode_single_layer.md`, and removing them from the public `deepseek_v4_decode_moe_router` signature was not the right tradeoff.

Pushed e5278f8 to restore the public entrypoint and `TensorSpec` inputs so the router still matches the single-layer/module integration contract. The internal learned-score inline kernel still omits those tensors because this branch does not consume them; the public function now documents that they are intentionally retained for hash-routing compatibility.

Validation after restoring the interface:
- `pre-commit run --show-diff-on-failure --color=always --all-files`
- `pytest tests/golden -q`
- `PYTHONPATH=$PWD PTOAS_ROOT=/home/bot/code/PTO/ptoas-bin PTO_ISA_ROOT=/home/bot/code/PTO/pto-isa ./.venv/bin/python models/deepseek/v4/deepseek_v4_decode_moe_router_draft.py -p a2a3sim`

File: models/deepseek/v4/deepseek_v4_decode_moe_router.py [hunk 1]
```diff
@@ -38,34 +40,210 @@
...
+    x_mixed:      pl.Tensor[[B, S, D],                    pl.BF16],
+    norm_w:       pl.Tensor[[D],                           pl.FP32],
+    gate_w:       pl.Tensor[[N_EXPERTS, D],                pl.FP32],
+    gate_bias:    pl.Tensor[[N_EXPERTS],                   pl.FP32],
+    tid2eid:      pl.Tensor[[VOCAB, TOPK],                 pl.INT32],
+    input_ids:    pl.Tensor[[B, S],                        pl.INT64],
+    x_norm:       pl.Tensor[[T, D],                        pl.BF16],
+    indices:      pl.Tensor[[T, TOPK],                     pl.INT32],
+    weights:      pl.Tensor[[T, TOPK],                     pl.FP32],
+):
```
[the remark this answers is not in the document]
Comment (zhaozhaozz) [reply models/deepseek/v4/deepseek_v4_decode_moe_router.py:65]: Partially addressed in the final squashed commit. I agree these tensors should not be part of the internal learned-score inline kernel, so `_deepseek_v4_decode_moe_router_kernel` no longer accepts them. I kept `tid2eid` / `input_ids` on the public `deepseek_v4_decode_moe_router` signature and in `build_tensor_specs()` because `deepseek_v4_decode_single_layer.md` defines them as part of the router integration contract for hash-routing compatibility. The public call site now documents that they are intentionally unused by the learned-score branch.

File: models/deepseek/v4/deepseek_v4_decode_moe_router_draft.py [hunk 2]
```diff
@@ -38,34 +40,210 @@
...
+    topk_idx_flat = pl.reshape(topk_idx_pad, [T * SCORE_PAD])
+    weight_out_flat = pl.reshape(weight_out_pad, [T * SCORE_PAD])
+    with pl.at(level=pl.Level.CORE_GROUP, name_hint="write_route_outputs"):
+        for t in pl.unroll(T):
+            dst_base = t * TOPK
+            src_base = t * SCORE_PAD
+            indices_flat = pl.write(indices_flat, [dst_base], pl.read(topk_idx_flat, [src_base]))
+            indices_flat = pl.write(indices_flat, [dst_base + 1], pl.read(topk_idx_flat, [src_base + 1]))
+            weights_flat = pl.write(weights_flat, [dst_base], pl.read(weight_out_flat, [src_base]))
+            weights_flat = pl.write(weights_flat, [dst_base + 1], pl.read(weight_out_flat, [src_base + 1]))
```
[the remark this answers is not in the document]
Comment (zhaozhaozz) [reply models/deepseek/v4/deepseek_v4_decode_moe_router_draft.py]: Addressed in the final squashed commit. The output scatter now uses `for k in pl.unroll(TOPK)` and writes all `TOPK` index/weight slots instead of hardcoding two writes.

File: models/deepseek/v4/deepseek_v4_decode_moe_router.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhaozhaozz) [reply models/deepseek/v4/deepseek_v4_decode_moe_router.py:65]: Not implementing the hash-routing branch in this PR. The current implementation targets the learned-score path (`N_HASH_LAYERS = 0` in this draft). Hash routing has different semantics: it should read `tid2eid[input_ids]` rather than compute gate scores/topk. I kept the public router signature compatible with the single-layer contract, but the actual hash branch should be added and validated as a separate change.

File: models/deepseek/v4/deepseek_v4_decode_moe_router_draft.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhaozhaozz) [reply models/deepseek/v4/deepseek_v4_decode_moe_router_draft.py]: Addressed in the final squashed commit. The router output scatter now loops over `pl.unroll(TOPK)` and updates `indices_flat` / `weights_flat` for every slot, so it is no longer tied to `TOPK == 2`.

File: models/deepseek/v4/deepseek_v4_decode_moe_router_draft.py [hunk 3]
```diff
@@ -38,34 +40,210 @@
...
+                w_row = pl.slice(gate_w, [1, D_CHUNK], [expert_i, d0])
+                prod = pl.col_expand_mul(x_chunk, w_row)
+                score_acc = pl.add(
+                    score_acc,
+                    pl.reshape(pl.row_sum(prod), [1, T]),
+                )
+            score_acc_buf = pl.assemble(score_acc_buf, score_acc, [0, 0])
+            bias = pl.read(gate_bias, [expert_i])
+            logits = pl.load(score_acc_buf, [0, 0], [1, T])
+            softplus = pl.log(pl.add(pl.exp(logits), 1.0))
```
[the remark this answers is not in the document]
Comment (zhaozhaozz) [reply models/deepseek/v4/deepseek_v4_decode_moe_router_draft.py]: Addressed in the final squashed commit. `softplus` now uses the stable form `max(x, 0) + log(1 + exp(-abs(x)))`, avoiding overflow for large positive logits while preserving the same mathematical result.
